validate_presence_of works with delegation

maurogeorge opened the pull request:
If we try define a validation to a attribute defined on a delegated object we
will receive a `NoMethodError` if the delegated attribute expect a collection as
value and not `nil` to a empty value.

Change the definition of what is a collection, before we are checking on
`:has_many` and `:has_and_belongs_to_many` association, now we check only if the
default value of the object is a empty array, since this is the default behavior
of a empty collection on `ActiveRecord`.

closes #765

mcmire commented on spec/unit/shoulda/matchers/active_model/validate_presence_of_matcher_spec.rb:
```diff
@@ -52,6 +52,42 @@
     end
   end
 
+  context "a ActiveModel that's validate a delegated attribute to a ActiveModel" do
```
What about:

"an ActiveModel with a validation on an attribute that delegates to an association in another ActiveModel"

mcmire commented on spec/unit/shoulda/matchers/active_model/validate_presence_of_matcher_spec.rb:
```diff
@@ -52,6 +52,42 @@
+    end
+  end
+
+  context "a ActiveModel that's validate a delegated attribute to a has_and_belongs_to_many association" do
```
What about:

"an ActiveModel with a validation on an attribute that delegates to a habtm association in another ActiveModel"

mcmire commented on spec/unit/shoulda/matchers/active_model/validate_presence_of_matcher_spec.rb:
```diff
@@ -52,6 +52,42 @@
+      @model = define_active_model_class('Example', accessors: [:parent]) do
+        delegate :children, :children=, to: :parent
+        validates_presence_of :children
+      end.new(parent: parent)
```
Can you put the `.new(parent: parent)` on its own line? This looks a little funny.

mcmire commented on spec/unit/shoulda/matchers/active_model/validate_presence_of_matcher_spec.rb:
```diff
@@ -52,6 +52,42 @@
+      parent = define_model :parent do
+        has_and_belongs_to_many :children
+        validates_presence_of :children
+      end.new
```
Can you put `.new` here on its own line?

Also there should be line breaks around methods that take blocks.

mcmire commented on spec/unit/shoulda/matchers/active_model/validate_presence_of_matcher_spec.rb:
```diff
@@ -52,6 +52,42 @@
+      @model = define_active_model_class('Example', accessors: [:parent]) do
+        delegate :children, :children=, to: :parent
+        validates_presence_of :children
+      end.new(parent: parent)
```
This `.new` should go on its own line, too.

maurogeorge commented:
@mcmire thanks for the review :heart_decoration:

I updated the code following your comments, and now the :dog: is barking should I follow the @houndci recomendation?

I updated other sections of the same file too :smile: I keep this on a separated commit since is a different change.

mcmire commented:
Oh I'm sorry by "place `.new` on its own line" I meant let's make another variable to hold the instance of the model on the next line.

maurogeorge commented:
@mcmire no problem :smile: 
I updated the code.

eparreno commented:
What happened with that fix? I having that issue with delegated methods.

mcmire commented:
Hey folks. In an effort to lighten our load as maintainers and be able to serve you better in the future, the shoulda-matchers team is working on cleaning out the cobwebs in this repo by pruning the backlog. As there are few of us, there are a lot of items that will simply never earn our attention in a reasonable time frame, and rather than giving you an empty promise, we think it makes more sense to focus on more recent issues. That means, unfortunately, that we must close this PR.

Don't take this the wrong way: our aim is not to diminish the effort people have made or dismiss problems that have been raised. If you feel that we should reopen this PR, then please let us know so that we can reprioritize it. Thanks!
